engine/src/test/java/net/jqwik/api/EMailsTests.java

```diff
@@ -216,6 +216,55 @@ void domainsWithOneAndMorePartsAreGenerated(@ForAll("emails") String email){
 
 	}
 
+	@Group
+	class ShrinkingTests {
+
+		@Property
+		void domainShrinking(@ForAll Random random){
+			Arbitrary<String> emails = Arbitraries.emails().filter(v -> domainShrinkingFilter(v));
+			assertAllValuesAreShrunkTo("a@a", emails, random);
+		}
+
+		@Property
+		void domainShrinkingWithTLD(@ForAll Random random){
+			Arbitrary<String> emails = Arbitraries.emails().filter(v -> domainShrinkingWithTLDFilter(v));
+			assertAllValuesAreShrunkTo("a@a.aa", emails, random);
+		}
```

Comment:
The runtime of this is very high (32s), the runtime of domainShrinking is also high (8s).
I think the problem is the generation of ip addresses. Do you know what I can do?

Reply:
Shrinking is a (comparatively) costly procedure. Especially with arbitraries with a lot of combinators, filters etc.

I suggest you reduce the number of tries for shrinking tests to 100 or even 10.

Reply:
The tries are not the problem as the tests are not working now (1 trie) and the runtime is so long

Reply:
One problem is that it shrinks (correctly) to "A@a.aa" because "A" is smaller (by asci code)  that "a"

Reply:
Another problem is 
```
private Arbitrary<String> domain(){
		return Arbitraries.frequencyOf(
				Tuple.of(1, domainIPv4()),
				Tuple.of(1, domainIPv6()),
				Tuple.of(2, domainDomain())
		);
	}
```
`Tuple.of(2, domainDomain())` should be the first option. Then it will be the preferred shrinking target.

Reply:
How can you change the shrinking order? And then reduce the runtime? Or should I change the testcases to "A@..."?

Reply:
One more thing. The filtering should (in this case) be through a falsifier, e.g.

```
@Property
void ipv4Shrinking(@ForAll Random random) {
	Arbitrary<String> emails = Arbitraries.emails();
	Falsifier<String> falsifier = email -> {
		String domain = getDomainOfEmail(email);
		if (isIPAddress(domain) && domain.contains("."))
			return TryExecutionResult.falsified(null);
		else
			return TryExecutionResult.satisfied();
	};
	String value = shrinkToMinimal(emails, random, falsifier);
	assertThat(value).isEqualTo("A@[0.0.0.0]");
}
```

This will greatly improve perfromance of tests